torch/fx/passes/runtime_assert.py

```diff
@@ -347,51 +388,547 @@ def go(node, keypath):
+                        # assert and also explicitly refine the range
+                        # (refinement should not be necessary once runtime
+                        # asserts cause refinement, but that's NYI)
+                        def convert(s):
+                            if s in (int_oo, -int_oo):
+                                return s
+                            try:
+                                return int(s)
+                            except TypeError:
+                                return None
+                        _refine_range(SymRel.GE, symbol, convert(vr.lower))
+                        _refine_range(SymRel.LE, symbol, convert(vr.upper))
+
+    # now handle existing runtime asserts in prioritized order, looking at is_size calls first
+    for node in is_size_calls + non_is_size_calls:
+        if node.target == torch.ops.aten._assert_async.msg:
+            expr_node = node.args[0].args[0]
+        else:
+            expr_node = node.args[0]
+
+        # trivial (simplified by dynamo if static), skip
+        if expr_node == True:
+            continue
+
+        log.debug("read runtime assert %s for node %s", node.target, node)
+
+        # maybe flip/negate expression
+        is_size = node.target in (
+            torch.ops.aten.sym_constrain_range_for_size.default,
+            torch._check_is_size,
+        )
+        expr = _get_sym_val(expr_node)
+        if not is_size:
+            old_expr = expr
+            expr = _canonicalize_expr(expr, sym_expr_to_ranges)
+            log.debug("canonicalize_expr(%s) -> %s", old_expr, expr)
+
+        # get signature, refine ranges
+        if node.target in (
+            torch.ops.aten._assert_scalar.default,
+            torch._check,
+            torch.ops.aten._assert_async.msg,
+        ):
+            if (sym_rel := _get_sym_rel(expr)) is None:
+                continue
+            sym_expr, val = expr.args
+            signature = (node.target, sym_rel, val)
+            # maybe we get a torch._check(u0 < s0), we expect this to show up in shape_env.deferred_runtime_asserts
+            # and handle it there instead.
+            if not isinstance(val, sympy.Number):
+                continue
+            _refine_range(sym_rel, sym_expr, val)
+
+        elif node.target in (
+            torch.ops.aten.sym_constrain_range_for_size.default,
+            torch.ops.aten.sym_constrain_range.default,
+        ):  # refine bounds, maybe mark size-like
+            _min = node.kwargs.get("min", 0 if node.target == is_size else -int_oo)
+            _max = node.kwargs.get("max", int_oo)
+            signature = (node.target, _min, _max)
+            _refine_range(SymRel.GE, expr, _min)
+            _refine_range(SymRel.LE, expr, _max)
+            sym_expr = expr
+
+        else:  # torch._check_is_size
+            signature = (node.target,)
+            sym_expr = expr
+
+        if is_size:  # make size-like
+            sym_expr_to_ranges[expr].set_is_size()
```

Comment:
All of this seems unnecessarily complicated. I think it would be better to do a version of this pass that is not trying to do nontrivial semantic deduplication. So if I had `u0 > 0` vs `u0 >= 1` as runtime asserts waiting in the ShapeEnv, I wouldn't try to deduplicate these. It's the responsibility for the ShapeEnv to remove duplicates.

So in this regime, there are only two things we need to track: is the symbol size-like (because that's special), and then otherwise a set of expressions which evaluate to True in `_assert_scalar`. We do not even need `sym_constrain_range`, because we can simply desugar this to `low <= u0` and `u0 <= upper`. And because we are CSE'ing, we get some mild semantic deduplication, which is that if I have an RA saying `low <= u0` and I also have a VR saying `low <= u0`, I will end up only generating the `_assert_scalar` once.